frontend/ui/widget/textboxwidget.lua

```diff
@@ -499,6 +500,25 @@ function TextBoxWidget:_splitToLines()
 
         ::idx_continue:: -- (Label for goto when use_xtext=true)
     end
+
+    if self.honor_soft_hyphen then
+        -- Check if there is a soft hyphen at the end of a line, replace it with a visible hyphen.
+        -- Begin from the end.
+        local soft_hyphens_at_eol = false
+        for i = #self.vertical_string_list, 1, -1 do
+            local pos = self.vertical_string_list[i].end_offset
+            if self.text:sub(pos, pos + 1) == "\xC2\xAD" then
+                self.text = self.text:sub(1, pos - 1) .. "-" .. self.text:sub(pos + 2)
```

Comment:
I initially thought you would have an issue with this way of doing it, by making the line of text bigger by replacing the 0-width soft hyphen with a real minus sign - that will make the line wider.
If the line was fit really near the target width, you could have that hyphen drawn outside the target width (so, in effect, you would get "hanging punctuation" that we all like :), as textboxwidget only extend spaces (for justification), and there's no code to reduce the width of spaces to make all the rest fit in the target width.

Which is not what will be happening because of what you do next. But what you do next causes a much bigger issue.